crates/autopilot/src/domain/settlement/trade/mod.rs

```diff
@@ -0,0 +1,247 @@
+use {
+    super::{transaction, transaction::Prices},
+    crate::domain::{
+        self,
+        auction::{self, order},
+        eth,
+        fee,
+    },
+    bigdecimal::Zero,
+    std::collections::HashSet,
+};
+
+mod math;
+
+/// Trade type when evaluated in a context of an Auction.
+#[derive(Clone, Debug)]
+pub enum Trade {
+    /// A regular user order. These orders are part of the `Auction`.
+    User(User),
+    /// A regular user order that was not part of the `Auction`.
+    UserOutOfAuction(User),
+    /// A JIT order that captures surplus. These orders are usually not part of
+    /// the `Auction`.
+    SurplusCapturingJit(Jit),
+    /// A JIT order that does not capture surplus, doesn't apply for protocol
+    /// fees and is filled at it's limit prices. These orders are never part of
+    /// the `Auction`.
+    Jit(Jit),
+}
```

Comment:
I think the enum variants are poorly named and chosen.

Structurally I see us - like in the driver domain - only having two types of trades
1. "normal" trades (called Fulfillment in the driver domain, why don't we stick to this term?)
2. jit orders

jit orders can be surplus capturing (which I think is best expressed via a bool on the `Jit` struct and not by giving this enum an completely different subtype.

`UserOutOfAuction` doesn't make sense to me and I don't see how this is used in practice in the protocol. I think they should just be considered non surplus capturing jit orders.

Reply:
`UserOutOfAuction` is used for user orders that exist in ordebook but were not part of an Auction. Although a violation, this can happen. I've explained somewhere that `Settlement` type needs to be built even for invalid settlements because even for invalid settlements (at least for the ones for which we can reconstruct the Auction) we need to save the breakdown of fees, jit orders etc.

Reply:
> which I think is best expressed via a bool on the Jit struct and not by giving this enum an completely different subtype

It can, although I am not sure that this is safe. For example, whenever we add a new function on `Trade` enum, we would need to remember that there is a possible different behaviour if the Jit is surplus capturing or not, which easily gets lost since the object has a lot of fields. Differently said, modelling two objects that behave completely differently in this way is not a good tradeof IMO just because they have a common property - not belonging to the orderbook.

Reply:
I don't buy this argument. The only thing that is different between surplus capturing jit order and regular jit orders is the score (which is taken care of in this PR and for which the distinction is hard to miss). Now we have an additional match arm in all other places, which is useless (and I don't see how it will be used in the future).

Why was this tradeoff then not considered in the driver domain? I'd be strongly in favor of keeping the two domains as similar as possible since the same engineers are maintaining both domains and shouldn't have to context switch for something that is expressing practically the same thing.

Reply:
> UserOutOfAuction is used for user orders that exist in ordebook but were not part of an Auction.

This concept doesn't exist in the protocol. Those are just regular jit orders. Please explain why they require their own top level type (it seems like they weren't handled until now just fine).

Reply:
> keeping the two domains as similar as possible

Agree, but it's a fact that driver and autopilot have differences. For example, driver can reject a solution containing fulfillment for an order that was not listed in the Auction, while autopilot must handle this case if it happens. 

> Why was this tradeoff then not considered in the driver domain?

From https://github.com/cowprotocol/services/pull/2764

> Since we are short on time we hackily convert surplus capturing JIT orders into Fulfillments (aka regular orders) when building the solution in the driver. That way they will get encoded the same way regular orders will.

We just never went back to rewrite this.

> This concept doesn't exist in the protocol. Those are just regular jit orders. Please explain why they require their own top level type (it seems like they weren't handled until now just fine).

One important difference is that circuit breaker is supposed to react if solver settles user order outside auction, if I am not wrong. Are you sure I shouldn't move these to be Jit and not Fulfilmment? We would then have the same order in `orders` and `jit_orders`. Maybe `jit_orders` shouldn't have order_uid as primary key but <block_number, log_index> then.

Reply:
I think we can just insert it into the jit order table. The circuit breaker, if they want to enforce this, can check both tables and deny-list a solver if it finds a user order in the jit order table. This way we don't convolute the autopilot domain with this somewhat esoteric edge case.